File: src/main/java/no/jstien/bikeapi/utils/holiday/HolidayRepositoryImpl.java
package no.jstien.bikeapi.utils.holiday;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import no.jstien.bikeapi.utils.CalendarUtils;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpUriRequest;
import org.apache.http.util.EntityUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.GregorianCalendar;
import java.util.List;

public class HolidayRepositoryImpl implements HolidayRepository {
    private static final String WEBAPI_NO_URL = "https://webapi.no";
    private static final String HOLIDAY_PATH = "/api/v1/holydays/";

    private static final Logger LOG = LogManager.getLogger();

    private HttpClient httpClient;
    private List<Holiday> holidays;
    private int cachedYear;

    public HolidayRepositoryImpl(HttpClient httpClient) {
        this.cachedYear = 0;
        this.holidays = Collections.emptyList();
        this.httpClient = httpClient;
    }

    @Override
    public List<Holiday> getHolidaysForCurrentYear() {
        // literally future proof
        if (cachedYear != getCurrentYear()) {
            refreshRegistry();
        }

        return Collections.unmodifiableList(holidays);
    }

    private void refreshRegistry() {
        try {
            HttpUriRequest request = createRequest();
            HttpResponse response = httpClient.execute(request);
            holidays = parseResponse(response);
            for (Holiday holiday : holidays) {
                LOG.debug("HOLIDAY: {} ({})", holiday.getDate(), holiday.getDescription());
            }
        } catch (Exception e) {
            LOG.error("Failed to refresh holiday registry", e);
        }
    }

    private HttpUriRequest createRequest() {
        return new HttpGet(WEBAPI_NO_URL + HOLIDAY_PATH + getCurrentYear());
    }

    private int getCurrentYear() {
        return ZonedDateTime.now().getYear();
    }

    private List<Holiday> parseResponse(HttpResponse response) {
        throwIfBadStatus(response);

        try {
            String responseJson = EntityUtils.toString(response.getEntity(), "UTF-8");
            return parseJson(responseJson);
        } catch (IOException e) {
            throw new RuntimeException("Failed to parse response JSON", e);
        }
    }

    private void throwIfBadStatus(HttpResponse response) {
        int statusCode = response.getStatusLine().getStatusCode();
        if (statusCode != 200) {
            throw new RuntimeException("Server returned status " + statusCode);
        }
    }

    private List<Holiday> parseJson(String json) {
        JsonParser parser = new JsonParser();
        JsonObject root = parser.parse(json).getAsJsonObject();


        JsonArray data = root.get("data").getAsJsonArray();
        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");

        List<Holiday> holidays = new ArrayList<>();
        data.forEach(ele -> {
            JsonObject obj = ele.getAsJsonObject();

            try {
                String desc = obj.get("description").getAsString();
                String dateStr = obj.get("date").getAsString();

                GregorianCalendar date = new GregorianCalendar();
                date.setTime(dateFormat.parse(dateStr));
                CalendarUtils.clearTime(date);

                holidays.add(new Holiday(date, desc));
            } catch (Exception e) {
                LOG.error("Failed to parse date", e);
            }
        });

        return holidays;
    }

}

File: src/main/java/no/jstien/bikeapi/utils/holiday/HolidayRepositoryImpl.kt
package no.jstien.bikeapi.utils.holiday

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import no.jstien.bikeapi.utils.CalendarUtils
import org.apache.http.HttpResponse
import org.apache.http.client.HttpClient
import org.apache.http.client.methods.HttpGet
import org.apache.http.client.methods.HttpUriRequest
import org.apache.http.util.EntityUtils
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

import java.io.IOException
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.time.ZonedDateTime
import java.util.ArrayList
import java.util.Collections
import java.util.GregorianCalendar

class HolidayRepositoryImpl(private val httpClient: HttpClient) : HolidayRepository {
    private var holidays: List<Holiday>? = null
    private val cachedYear: Int

    override// literally future proof
    val holidaysForCurrentYear: List<Holiday>
        get() {
            if (cachedYear != currentYear) {
                refreshRegistry()
            }

            return Collections.unmodifiableList(holidays!!)
        }

    private val currentYear: Int
        get() = ZonedDateTime.now().year

    init {
        this.cachedYear = 0
        this.holidays = emptyList()
    }

    private fun refreshRegistry() {
        try {
            val request = createRequest()
            val response = httpClient.execute(request)
            holidays = parseResponse(response)
            for (holiday in holidays!!) {
                LOG.debug("HOLIDAY: {} ({})", holiday.date, holiday.description)
            }
        } catch (e: Exception) {
            LOG.error("Failed to refresh holiday registry", e)
        }

    }

    private fun createRequest(): HttpUriRequest {
        return HttpGet(WEBAPI_NO_URL + HOLIDAY_PATH + currentYear)
    }

    private fun parseResponse(response: HttpResponse): List<Holiday> {
        throwIfBadStatus(response)

        try {
            val responseJson = EntityUtils.toString(response.entity, "UTF-8")
            return parseJson(responseJson)
        } catch (e: IOException) {
            throw RuntimeException("Failed to parse response JSON", e)
        }

    }

    private fun throwIfBadStatus(response: HttpResponse) {
        val statusCode = response.statusLine.statusCode
        if (statusCode != 200) {
            throw RuntimeException("Server returned status $statusCode")
        }
    }

    private fun parseJson(json: String): List<Holiday> {
        val parser = JsonParser()
        val root = parser.parse(json).asJsonObject


        val data = root.get("data").asJsonArray
        val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss")

        val holidays = ArrayList<Holiday>()
        data.forEach { ele ->
            val obj = ele.asJsonObject

            try {
                val desc = obj.get("description").asString
                val dateStr = obj.get("date").asString

                val date = GregorianCalendar()
                date.time = dateFormat.parse(dateStr)
                CalendarUtils.clearTime(date)

                holidays.add(Holiday(date, desc))
            } catch (e: Exception) {
                LOG.error("Failed to parse date", e)
            }
        }

        return holidays
    }

    companion object {
        private val WEBAPI_NO_URL = "https://webapi.no"
        private val HOLIDAY_PATH = "/api/v1/holydays/"

        private val LOG = LogManager.getLogger()
    }

}
